src/week3.hs

```diff
@@ -0,0 +1,172 @@
+module HW03 where
+
+data Expression =
+    Var String                   -- Variable
+  | Val Int                      -- Integer literal
+  | Op Expression Bop Expression -- Operation
+  deriving (Show, Eq)
+
+-- Binary (2-input) operators
+data Bop = 
+    Plus     
+  | Minus    
+  | Times    
+  | Divide   
+  | Gt
+  | Ge       
+  | Lt  
+  | Le
+  | Eql
+  deriving (Show, Eq)
+
+data Statement =
+    Assign   String     Expression
+  | Incr     String
+  | If       Expression Statement  Statement
+  | While    Expression Statement       
+  | For      Statement  Expression Statement Statement
+  | Sequence Statement  Statement        
+  | Skip
+  deriving (Show, Eq)
+
+type State = String -> Int
+
+-- Exercise 1 -----------------------------------------
+
+extend :: State -> String -> Int -> State
+extend oldState str val = state
+    where state k
+            |k == str = val
+            |otherwise = oldState k
```

Comment:
this is of course correct but if you wanted to avoid having to do a `where` clause and define state directly ad a function, you can use lambda abstraction :)

```suggestion
extend oldState str val = \k ->
  if k == str then val else oldState k
```

note: I changed to `if..then..else` style because unfortunately this [doesn't play as well with guards](https://stackoverflow.com/questions/3416475/haskell-guards-on-lambda-functions/3416514)